File: src/Python/Snake.py
import pygame

from Constants import *

class Snake:
	def __init__(self, surface):
		self.body = [[0, 0]]
		self.surface = surface
		self.speed = [0, 0]

	def draw(self):
		for position in self.body:
			pygame.draw.rect(self.surface, GREEN, pygame.Rect(position[0] * PIXEL_OFFSET, position[1] * PIXEL_OFFSET, PIXEL_OFFSET, PIXEL_OFFSET))

	def update(self):
		currentPosition = self.body[0].copy()

		currentPosition[0] += self.speed[0]
		currentPosition[1] += self.speed[1]

		if currentPosition[0] > GRID[0]:
			currentPosition[0] = 0
		elif currentPosition[0] < 0:
			currentPosition[0] = GRID[0]
		elif currentPosition[1] > GRID[1]:
			currentPosition[1] = 0
		elif currentPosition[1] < 0:
			currentPosition[1] = GRID[1]

		self.body.pop()
		self.body.insert(0, currentPosition)

	def addBody(self):
		self.body.append([self.body[-1][0] - self.speed[0], self.body[-1][1] - self.speed[1]])

	def eat(self, fruit):
		self.addBody()

		fruit.spawn(self)

	def moveLeft(self):
		self.speed = [-1, 0]

	def moveRight(self):
		self.speed = [1, 0]

	def moveUp(self):
		self.speed = [0, -1]

	def moveDown(self):
		self.speed = [0, 1]

	def hasCollidedWith(self, body):
		for position in body:
			if position == self.body[0]:
				return True

		return False

	def hasCollidedWithItself(self):
		return self.hasCollidedWith(self.body[1:])

	def hasReachedMaxLength(self):
		return len(self.body) == GRID[0] ** 2

File: src/Javascript/js/Snake.js
class Snake {
	constructor(ctx) {
		this.body = [[0, 0]];
		this.ctx = ctx;
	}

	addBody() {
		this.body.push(this.body[this.body.length - 1].map((pos, index) => pos - speed[index]));
	}

	hasCollidedWithFruit() {
		return this.body[0][0] == fruit.position[0] && this.body[0][1] == fruit.position[1];
	}

	hasCollidedWithItself() {
		return this.body.slice(1).some(pos => pos[0] == this.body[0][0] && pos[1] == this.body[0][1])
	}

	hasReachedMaximumBody() {
		return this.body.length == (GRID[0] - 1) ** 2;
	}

	calculateNextPosition() {
		return this.body[0].map((pos, index) => pos + speed[index]);
	}

	draw() {
		this.ctx.fillStyle = 'green';
			
		for (const position of this.body)
			ctx.fillRect(...position.map(pos => pos * PIXEL_OFFSET), PIXEL_OFFSET, PIXEL_OFFSET);
	}

	update() {
		// Update position
		let [x, y] = this.calculateNextPosition();
				
		if (x * PIXEL_OFFSET >= canvas.width) {
			x = 0;
		}
				
		if (x * PIXEL_OFFSET <= -PIXEL_OFFSET) {
			x = GRID[0];
		}
		
		if (y * PIXEL_OFFSET >= canvas.height) {
			y = 0;
		}
		
		if (y * PIXEL_OFFSET <= -PIXEL_OFFSET) {
			y = GRID[1];
		}
		
		this.body.pop();
		this.body.unshift([x, y]);

		// Has Collided with Fruit
		if (this.hasCollidedWithFruit())
			this.eat();

		// Update Game Status
		if (this.hasCollidedWithItself()) {
			status = 'lose';
		} else if (this.hasReachedMaximumBody()) {
			status = 'win';
		}
	}

	eat() {
		this.addBody();
		fruit.spawn();
	}
}